expression/builtin_cast.go

```diff
@@ -1830,7 +1831,10 @@ func BuildCastCollationFunction(ctx sessionctx.Context, expr Expression, ec *Exp
 	if expr.GetType().EvalType() != types.ETString {
 		return expr
 	}
-	if expr.GetType().GetCollate() == ec.Collation {
+	// Avoid padding 0 when cast character string to binary,
+	// which might affect string comparision result.
+	// See https://github.com/pingcap/tidb/issues/34823 for details.
+	if expr.GetType().GetCollate() == ec.Collation || ec.Charset == charset.CharsetBin {
```

Comment:
we can not just return if the target charset is `binary`. 

because we have had support charset `gbk`, if charset of `t1.a` is `gbk` it will have a `to_binary` function, after we return directly, it will be missed I think.

```sql
mysql> desc select * from t1, t2 where t1.a between t2.b and t2.c;
+-----------------------------+---------+-----------+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------+
| id                          | estRows | task      | access object | operator info                                                                                                                                 |
+-----------------------------+---------+-----------+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------+
| HashJoin_8                  | 4.00    | root      |               | CARTESIAN inner join, other cond:ge(cast(to_binary(test.t1.a), binary(20)), test.t2.b), le(cast(to_binary(test.t1.a), binary(20)), test.t2.c) |
| ├─TableReader_13(Build)     | 2.00    | root      |               | data:TableFullScan_12                                                                                                                         |
| │ └─TableFullScan_12        | 2.00    | cop[tikv] | table:t2      | keep order:false, stats:pseudo                                                                                                                |
| └─TableReader_11(Probe)     | 2.00    | root      |               | data:TableFullScan_10                                                                                                                         |
|   └─TableFullScan_10        | 2.00    | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                                                                                                |
+-----------------------------+---------+-----------+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------+
```

Reply:
``` sql
create table t4(b binary(20), c char(20)) collate utf8mb4_general_ci;
create table t3(a char(20) ) collate utf8mb4_general_ci;
insert into t3 values ("a");
insert into t4 values (0x0, "A");
select * from t3, t4 where t3.a between t4.b and t4.c;
```
The result is wrong because t3.a doesn't convert to binary collation.

Reply:
For `gbk`, tidb will build a `to_binary` is because `gbk` doesn't belong to legacy charset and when argument is legacy charset (for example the default charset utf8mb4_bin), there is no need to build `to_binary` function. [HandleBinaryLiteral](https://github.com/pingcap/tidb/blob/fb67ee441281b42032cf38d90ca726f3fdd7545f/expression/builtin_convert_charset.go#L309).

My change actually will still add the `to_binary` function when the code build function for `le/ge`.

```mysql
mysql> drop table t1;
Query OK, 0 rows affected (0.01 sec)

mysql> create table t1(a char(20)) default charset gbk;
Query OK, 0 rows affected (0.00 sec)

mysql> desc select * from t1, t2 where t1.a between t2.b and t2.c;
+-------------------------------+----------+-----------+---------------+-----------------------------------------------------------------------------------------------------------+
| id                            | estRows  | task      | access object | operator info                                                                                             |
+-------------------------------+----------+-----------+---------------+-----------------------------------------------------------------------------------------------------------+
| Projection_9                  | 20000.00 | root      |               | test.t1.a, test.t2.b, test.t2.c                                                                           |
| └─HashJoin_11                 | 20000.00 | root      |               | CARTESIAN inner join, other cond:ge(to_binary(test.t1.a), test.t2.b), le(to_binary(test.t1.a), test.t2.c) |
|   ├─TableReader_13(Build)     | 2.00     | root      |               | data:TableFullScan_12                                                                                     |
|   │ └─TableFullScan_12        | 2.00     | cop[tikv] | table:t2      | keep order:false, stats:pseudo                                                                            |
|   └─TableReader_15(Probe)     | 10000.00 | root      |               | data:TableFullScan_14                                                                                     |
|     └─TableFullScan_14        | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                                                            |
+-------------------------------+----------+-----------+---------------+-----------------------------------------------------------------------------------------------------------+
6 rows in set (0.01 sec)
```

Reply:
As for 

```mysql
create table t4(b binary(20), c char(20)) collate utf8mb4_general_ci;
create table t3(a char(20) ) collate utf8mb4_general_ci;
insert into t3 values ("a");
insert into t4 values (0x0, "A");
select * from t3, t4 where t3.a between t4.b and t4.c;
```

It is the expected behavior that t3.a doesn't convert to binary collation since its charset (utf8mb4) is legacy...

Reply:
> As for
> 
> ```sql
> create table t4(b binary(20), c char(20)) collate utf8mb4_general_ci;
> create table t3(a char(20) ) collate utf8mb4_general_ci;
> insert into t3 values ("a");
> insert into t4 values (0x0, "A");
> select * from t3, t4 where t3.a between t4.b and t4.c;
> ```
> 
> It is the expected behavior that t3.a doesn't convert to binary collation since its charset (utf8mb4) is legacy...

The point is that we need to make t3.a <= t4.c use the binary collation.

Reply:
Will `to_binary` padding 0 for other charsets? If so, then forcing t3.a <= t4.c to use the binary collation will cause mismatch between tidb and mysql. And mysql actually doesn't do conversion when target charset is `binary`.